packages/freezed/lib/src/freezed_generator.dart

```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
   }
 
   List<Parameter> _commonParametersBetweenAllConstructors(
-    List<ConstructorDetails> constructorsNeedsGeneration,
-  ) {
+    List<ConstructorDetails> constructorsNeedsGeneration, {
+    required bool allowCommonSuperType,
+  }) {
     return constructorsNeedsGeneration.first.parameters.allParameters
         .map((parameter) {
-          var hasAnyFinalProperty = false;
-          for (final constructor in constructorsNeedsGeneration) {
-            final matchingParameter =
-                constructor.parameters.allParameters.firstWhereOrNull((p) {
-              return p.name == parameter.name && p.type == parameter.type;
-            });
+          var anyMatchingPropertyIsFinal = false;
+          var isCommonWithDifferentNullability = false;
+
+          DartType? commonType;
+
+          for (final constructor in constructorsNeedsGeneration.skip(1)) {
+            final matchingParameter = constructor.parameters.allParameters
+                .firstWhereOrNull((p) => p.name == parameter.name);
 
             if (matchingParameter == null) return null;
-            if (matchingParameter.isFinal) hasAnyFinalProperty = true;
-          }
 
-          if (hasAnyFinalProperty) {
-            return parameter.copyWith(isFinal: true);
+            final commonTypeWithMatch =
```

Comment:
So, if I'm reading this correctly, this code repeatedly checks the least upper bound for the parameter in the first constructor and the parameter in another constructor. Only if this always returns the same result it will be considered.

I think this could be improved by declaring a `commonType` outside of the `for` loop, initializing it to the first parameter type, and then in the loop for each parameter type updating the `commonType` to the least upper bound between the `commonType` so far and the current parameter type.

That way the least upper bound between all parameter types will be found, even if that is different from the least upper bound between the first constructor and the second. Consider this example:

```dart
class A {}
class B implements A {}
class C implements B {}
class D implements A {}

@freezed
class Container {
  const factory Container.one({required B param}) = _One;
  const factory Container.two({required C param}) = _Two;
  const factory Container.three({required D param}) = _Three;
}
```
The types `B`, `C` and `D` are all subclasses of `A`, but currently this is not recognized.

Reply:
Tim already pointed that out. I have added a test for your use case and the parameter is available as expected.

Reply:
Sorry, I missed that.